tests/core/abi-utils/test_abi_utils.py

```diff
@@ -173,3 +694,440 @@ def test_event_signature_to_log_topic(event_signature, expected):
+            "uint256",
+            "uint256",
+        ),
+    ),
+)
+def test_collapse_if_tuple(abi_component, expected):
+    assert collapse_if_tuple(abi_component) == expected
+
+
+@pytest.mark.parametrize(
+    "abi_component",
+    (
+        (
+            {
+                "type": False,
+            }
+        ),
+    ),
+)
+def test_collapse_if_tuple_with_errors(abi_component):
+    with pytest.raises(TypeError):
+        collapse_if_tuple(abi_component)
+
+
+@pytest.mark.parametrize(
+    "fn_abi,args,kwargs,expected",
+    (
+        (
+            FN_ABI_C,
+            (
+                "0x1234567890123456789012345678901234567890",
+                b"bar",
+                "0x1234567890123456789012345678901234567890",
+            ),
+            {},
+            (
+                "0x1234567890123456789012345678901234567890",
+                b"bar",
+                "0x1234567890123456789012345678901234567890",
+            ),
+        ),
+        (
+            FN_ABI_C,
+            [],
+            {
+                "a": "0x1234567890123456789012345678901234567890",
+                "b": b"bar",
+                "c": "0x1234567890123456789012345678901234567890",
+            },
+            (
+                "0x1234567890123456789012345678901234567890",
+                b"bar",
+                "0x1234567890123456789012345678901234567890",
+            ),
+        ),
+    ),
+)
+def test_get_normalized_abi_inputs(fn_abi, args, kwargs, expected):
+    inputs = get_normalized_abi_inputs(fn_abi, args, kwargs)
+
+    assert inputs == expected
+
+
+def test_get_normalized_abi_inputs_returns_args_without_kwargs():
+    args = [1, "0x1234567890123456789012345678901234567890", b"bar"]
+    kwargs = {}
+    inputs = get_normalized_abi_inputs(FN_ABI_C, args, kwargs)
+
+    assert inputs == args
+
+
+def test_error_when_invalid_args_kwargs_combo_provided():
+    with pytest.raises(TypeError, match="Incorrect argument count."):
+        get_normalized_abi_inputs(
+            FN_ABI_E,
+            (
+                1,
+                2,
+            ),
+            {"a": 1, "b": 2},
+        )
+
+
+def test_kwargs_is_disallowed_when_merging_with_unnamed_inputs():
+    with pytest.raises(TypeError):
+        get_normalized_abi_inputs(FN_ABI_E, tuple(), {"x": 1, "y": 2})
+
+
+def test_args_works_when_merging_with_unnamed_inputs():
+    actual = get_normalized_abi_inputs(FN_ABI_E, (1, 2), {})
+    assert actual == (1, 2)
+
+
+def test_args_allowed_when_duplicate_named_inputs():
+    actual = get_normalized_abi_inputs(FN_ABI_DUPLICATE_NAMES, (1, 2, 3), {})
+    assert actual == (1, 2, 3)
+
+
+def test_kwargs_not_allowed_for_duplicate_input_names():
+    with pytest.raises(
+        TypeError, match=re.escape("testFn() got multiple values for argument(s) 'b'")
+    ):
+        get_normalized_abi_inputs(FN_ABI_DUPLICATE_NAMES, (1,), {"a": 2, "b": 3})
+
+
+def test_kwargs_allowed_if_no_intersections_with_duplicate_input_names():
```

Comment:
I'm not sure I understand this test case: "kwargs allowed if" yet it raises an exception. Is this test correct? It seems like it should allow this exact test case base on the test name. Which expectation is it?

Reply:
This is certainly confusing. Now that you flagged this, I am confused by a couple of cases.
Components having inputs with the same names should be invalid.

```python
>>> FN_ABI_DUPLICATE_NAMES = {
...     "constant": False,
...     "inputs": [
...         {"name": "b", "type": "int256"},
...         {"name": "b", "type": "int256"},
...         {"name": "a", "type": "int256"},
...     ],
...     "name": "testFn",
...     "outputs": [],
...     "type": "function",
... }

>>> get_normalized_abi_inputs(FN_ABI_DUPLICATE_NAMES, (1, 2, 3), {})
(1, 2, 3)

>>> get_normalized_abi_inputs(FN_ABI_DUPLICATE_NAMES, (1,), {"a": 2, "b": 3})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stu/projects/ethereum/eth-utils/eth_utils/abi.py", line 449, in get_normalized_abi_inputs
    raise TypeError(
TypeError: testFn() got multiple values for argument(s) 'b'.
```


 This existed in the private function, but I believe inputs can have the same names (especially for events). https://github.com/ethereum/web3.py/blob/a721a41e669b2ee58afa2c0c1061486bd0041802/web3/_utils/abi.py#L415